GH-1081 Auto-dismiss notification.

lmcbout opened the pull request:
Auto-dismiss the notification
Increase the timer when the user has choices to decide

Signed-off-by: Jacques Bouthillier <jacques.bouthillier@ericsson.com>

marcdumais-work commented:
I think the original idea was to have the auto-dismiss delay be optional (and I guess optionally not dismiss at all). See https://github.com/theia-ide/theia/issues/1081

We can continue the discussion in that issue?

epatpol commented on packages/messages/src/browser/notifications.ts:
```diff
@@ -72,8 +74,19 @@ export class Notifications {
+            let adjustedCloseTimer = this.defaultAutoCloseTimeout;
+            // Give more time before activating the autoClose Notification when user has choices
+            if (properties.actions !== undefined && properties.actions.length > 1) {
+                adjustedCloseTimer = adjustedCloseTimer * properties.actions.length;
```
Just for reference, we discussed this offline and the conclusion is that it's better to make it auto-dismissable for certain severity level i.e info/warning and make it must-be-dismissed-manually for things like errors. 

Also I don't think using the number of possible actions in the message is a good factor to use for the duration of the message, as there's probably no correlation between the number of actions a user can do on that message and the severity of it.

AlexTugarev replied:
I overlooked your comment, as already collapsed.

That's the thing, all kind of use cases require different options. I think, it's up to the `MessageService` clients to decide about timeout and other properties of messages. OTOH we can have default options for most common cases.

AlexTugarev commented on packages/messages/src/browser/notifications.ts:
```diff
@@ -72,8 +74,18 @@ export class Notifications {
             }
+
+            // Do not auto-close the "error" notification
+            if (properties.icon.toLowerCase() !== 'error') {
```
it should be configurable via properties of a `Message` as required in  https://github.com/theia-ide/theia/issues/1081#issue-289915333, because the client needs to decide, if this message should be sticky or not.

lmcbout commented:
I Have a question:
I put the auto-Dismiss value optional in message-service.ts.
When I call the service in window-impl.ts or in cpp-client-contribution, I needed to add an extra parameter even if it was optional. I put "undefined" for the auto-dismiss value even if there was other optionals value defined. When I tested with task, I did not need to define the optional value since there was nothing else than the message text. I was able to add an auto-dismiss value  and it was working fine.
Is there a way to have a second optional fields and not put "undefined" if we don't plan to use a timer for the auto-dismiss notification ?

AlexTugarev commented on packages/languages/src/common/window-impl.ts:
```diff
@@ -28,16 +28,16 @@ export class WindowImpl implements Window {
         };
         if (type === MessageType.Error) {
-            return this.messageService.error(message, ...actionTitles).then(mapActionType);
+            return this.messageService.error(message, undefined, ...actionTitles).then(mapActionType);
```
we should introduce function overload to keep the less verbose API for most common use cases.

lmcbout replied:
Overload method added, so no need to use undefined anymore

AlexTugarev commented on packages/messages/src/browser/notification-preferences.ts:
```diff
@@ -0,0 +1,46 @@
+export const NotificationConfigSchema: PreferenceSchema = {
+    "type": "object",
+    "properties": {
+        "notification.autoDismiss": {
```
please rename it to something like `notification.timeout`. I'm planing to revise more parts of user notifications in #1360, and I think we should distinct between dismissing and hiding a message. a timeout is  more generics, which leaves more room for interpretation.

AlexTugarev replied:
just to mention it, hiding message and keeping them in kind of an inbox are ideas from #1360. I'm aware of this PR to target the closing of messages after the timeout period elapsed.

lmcbout replied:
auto-dismiss rename to timeout

AlexTugarev commented on packages/cpp/src/browser/cpp-client-contribution.ts:
```diff
@@ -45,7 +45,7 @@ export class CppClientContribution extends BaseLanguageClientContribution {
                 "Please make sure 'clangd' is installed on your system. " +
                 "You can refer to the clangd page for instructions.";
-            this.messageService.error(ERROR_MESSAGE, READ_INSTRUCTIONS_ACTION).then(selected => {
+            this.messageService.error(ERROR_MESSAGE, undefined, READ_INSTRUCTIONS_ACTION).then(selected => {
```
you see, it's inconvenient to add this `undefined` parameter all over the places.

lmcbout replied:
agree, no more needed to use undefined with the overload method

AlexTugarev commented on packages/core/src/common/message-service.ts:
```diff
@@ -15,20 +15,30 @@ export class MessageService {
+     * If we set autoDismiss: 0, then the notification preference will be ignored and notification remains until user intervention.
+     */
+
+    log(message: string, autoDismiss?: number | undefined, ...actions: string[]): Promise<string | undefined> {
```
also we should add `MessageOptions` to encapsulate, as we might add more of them.

epatpol replied:
@AlexTugarev what about also putting actions in that MessageOptions structure?

AlexTugarev replied:
Good question. Should the options describe the message, or it's rendering/behavior. 
I'd also like to improve `actions` of messages, as they currently are label and value at the same time. Have a look at the resolution of resulting promises, that's what you selected. 
OTOH the very first and simplest signature from above is quiet handy, and should be kept. 
But do you think, we need to tackle that point now?

epatpol replied:
Oh sorry I didn't see the message above where you mentionned function overloads. Yes that could work also, I just thought that if they were also options they could go in the same structure, but indeed it looks like the structure might not be appropriate for them.

AlexTugarev commented on packages/messages/src/browser/notifications-message-client.ts:
```diff
@@ -35,14 +41,17 @@ export class NotificationsMessageClient extends MessageClient {
             fn: element => onCloseFn(action)
         });
+
+        const timeout = (message.timeout || 0);
```
if `message.actions` are undefined or empty, timeout should be undefined, right?

AlexTugarev replied:
you're right! the existence of actions should be the first criteria.

but I think instead of patching the timeout in `NotificationsMessageClient.showMessage` and evaluating everything in `Notifications.contructor`, we should compute the effective timeout here onve. considering the requirement could we calculate the timeout like this?
```
const timeout = (message.action)
  ? undefined
  : (message.timeout || this.preferences['notification.timeout'])
```

could you update the PR, please?

AlexTugarev commented on packages/messages/src/browser/notifications.ts:
```diff
@@ -72,8 +74,19 @@ export class Notifications {
+                    clearTimeout(this.closeTimer);
                 });
             }
+            // Auto-dismiss is only allowed when there is only one action available.
```
please move the the check to `showToast`, it will simplify things here.

AlexTugarev commented on packages/messages/src/browser/notifications.ts:
```diff
@@ -29,6 +30,7 @@ export interface Notification {
 
 export class Notifications {
 
+    private closeTimer: any;
```
it's `number | undefined` if you use `window.setTimeout` below.

AlexTugarev replied:
1. I think this should go into the scope of `createNotificationElement`
2. we can avoid use `any` here, as already commented

lmcbout replied:
Move the closeTimer to createNotificationElement

lmcbout commented:
ready to be reviewed

AlexTugarev commented on packages/messages/src/browser/notifications.ts:
```diff
@@ -72,8 +74,16 @@ export class Notifications {
                 button.addEventListener('click', () => {
                     action.fn(handler);
                     close();
+                    clearTimeout(this.closeTimer);
```
@lmcbout here you're referencing the `closeTimer` property of the `Notifications` instance. this is not, what you want to say it should do, right? you want to clear the timeout of the single message, but this way each click handler will clear the timeout of the latest shown message. I think I already commented on this, but GH comments were rooted out.

lmcbout replied:
Craete a clearTimeout() and setTimeout, so I can use the window.setTimeout and window.cearTimeout() method

lmcbout replied:
No need to create the set and clear timeout method, using wndow.set instead

AlexTugarev commented on packages/messages/src/browser/notifications.ts:
```diff
@@ -72,8 +74,16 @@ export class Notifications {
+                    clearTimeout(this.closeTimer);
                 });
             }
+            if (properties.timeout !== undefined && properties.timeout > 0) {
```
minor: `!!properties.timeout && ...` is shorter

lmcbout replied:
Done this minor modification

AlexTugarev commented on packages/messages/src/browser/notifications.ts:
```diff
@@ -72,8 +74,16 @@ export class Notifications {
                 button.addEventListener('click', () => {
```
shouldn't we clear the timeout before `fn` call?

lmcbout replied:
Should not matter, but I will mve the clearTimeout before the close() call

AlexTugarev commented on packages/messages/src/browser/notifications.ts:
```diff
@@ -72,8 +74,16 @@ export class Notifications {
+                    clearTimeout(this.closeTimer);
                 });
             }
+            if (properties.timeout !== undefined && properties.timeout > 0) {
```
you can move this `if` statement up before L70 to make a local `closeTimer`. 
something like  `const closeTimer = !!properties.timeout ? window.setTimeout() : undefined` should work.
then you can use it in click handler.

lmcbout replied:
Move the closeTimer before and use it as a const

AlexTugarev commented on packages/messages/src/browser/notifications.ts:
```diff
@@ -72,8 +74,16 @@ export class Notifications {
             }
+            if (properties.timeout !== undefined && properties.timeout > 0) {
+                this.closeTimer = setTimeout(() => {
+                    if (this.closeTimer !== undefined) {
```
`if` won't be necessary then.

lmcbout replied:
Removed

AlexTugarev commented on packages/messages/src/browser/notifications-message-client.ts:
```diff
@@ -35,14 +37,24 @@ export class NotificationsMessageClient extends MessageClient {
             fn: element => onCloseFn(action)
         });
+
+        // If some actions are defined, we don't set the timeout
```
convert to JSDoc, please.

lmcbout replied:
Using macro does not need anymore the comment

AlexTugarev commented on packages/core/src/common/message-service-protocol.ts:
```diff
@@ -21,12 +21,17 @@ export interface Message {
     type: MessageType;
     text: string;
     actions?: string[];
+    timeout?: number;
```
shouldn't that be a `options?: MessageOptions` property?

lmcbout replied:
Will be name options: MessageOptions

AlexTugarev commented on packages/core/src/common/message-service.ts:
```diff
@@ -6,29 +6,71 @@
     ) { }
+    /**
+     *
+     * MessageSevice uses some methods to report information either log, info, warn and error
```
please, convert to JSDoc.

AlexTugarev commented on packages/core/src/common/message-service.ts:
```diff
@@ -6,29 +6,71 @@
+    /**
+     *
+     * MessageSevice uses some methods to report information either log, info, warn and error
+     * One parameter is optional:  @param options.timeout
```
this doc should go to MessageOptions, right?

AlexTugarev commented on packages/core/src/common/message-service.ts:
```diff
@@ -6,29 +6,71 @@
+        let actionsValue: string[] | undefined;
+
+        if (args && args.length > 0) {
+            if (args[0]['timeout'] !== undefined) {
```
please use [type guards](https://www.typescriptlang.org/docs/handbook/advanced-types.html) if possible. also, I think we can assume `args[0]` to be message options if type is object.

AlexTugarev replied:
would be nice, if you add the type guard in the namespace of `MessageOptions`.

lmcbout replied:
In the next version, I remove theoverloa method by addibg the actions and timeout within the same API, so the typeguard are not require, the code is simplier here

AlexTugarev commented on packages/core/src/common/message-service.ts:
```diff
@@ -6,29 +6,71 @@
+            } else {
+                actionsValue = args;
+            }
+            return this.client.showMessage({ type: messageType, timeout: timeoutvalue, text: messageText, actions: actionsValue });
```
there is no need to explode the properties of options here.

AlexTugarev commented on packages/core/src/common/message-service.ts:
```diff
@@ -6,29 +6,71 @@
-    log(message: string, ...actions: string[]): Promise<string | undefined> {
-        return this.client.showMessage({ type: MessageType.Log, text: message, actions: actions || [] });
+    log(message: string, ...actions: string[]): Promise<string | undefined>;
+    log(message: string, options?: MessageOptions): Promise<string | undefined>;
```
is this one actually needed? please check, if it's same as the next line.

lmcbout commented:
With the commit : fd4b468
In file message-service.ts, I added a new API structure for the message(MessageArgs). I tested with the error method only.
With this API structure, it allow us to add any number of "optional " fields or structure. It prevent us to create overload method. Long term, it would be better. Short term, the drawback we need to adjust all messageService being used. Also, when using the messageService like :
OLD: this.messageService.error(`Error launching task '${taskName}': ${error}` );
instead of:
NEW: this.messageService.error({ message: `Error launching task '${taskName}': ${error}`, options: { timeout: 1357 } });

we need to use the API data structure ({message:...}) and we can add optional data as well

AlexTugarev commented:
@lmcbout, we discussed that kind of change before, and decided to have simple parameters to be less verbose. just compare the two examples, and consider that a lot of use cases won't require anythings but the message.

also, this is not about us avoid overload methods for MessageService, but to provide a decent set of methods for API clients to cover different use cases and avoid unnecessary structures.

I don't see the need to break the `MessageService` API. `MessageOptions` are optional and give us enough space to add more options in future. please have a look at the usages of the `MessageService`, what kind of issues would that change solve?

lmcbout commented:
Adjusted the MessageOption API to include the optional timeout and actions within the same API. This allowed me to simplify the code and allow the end-user to enter just a message string if no actions or timeout were added. Refering to some link on the web, it seems the way I implemented the optional parameter is the right one. Also, all web sites were in favor not to use to much the overload in Typescript.
Code ready to be reviewed

AlexTugarev commented:
@lmcbout 
> Refering to some link on the web, it seems the way I implemented the optional parameter is the right one. 

Which sources?

> Also, all web sites were in favor not to use to much the overload in Typescript.

Sources? That sounds a bit arbitrary for an explanation. Are there technical reasons? 
To emphasize this use case, let's focus on adding new options without breaking API, when there are no advantages gained.

AlexTugarev commented on packages/core/src/common/message-service-protocol.ts:
```diff
@@ -20,13 +20,18 @@ export enum MessageType {
+    options?: MessageOptions;
+}
+
+export interface MessageOptions {
```
the options were meant to add more hints, so that `MessageClient`s could decide on how to handle them. 

please, move the actions back to `Message`.

AlexTugarev replied:
should be `message.actions`, **not** `message.options!.actions`.

AlexTugarev commented:
@lmcbout please, reintroduce the overload methods for `MessageService`. 

https://github.com/theia-ide/theia/blob/dd4bc3f35cb81702c96eb34b95c9661acb6494c9/packages/languages/src/common/window-impl.ts#L30-L32

such a change does require to adapt all caller sites, but there is no noticeable advantage. I think we should keep this convenience API for actions at the moment. 

also, as stated before, we can consider to revise the actions. but this should be in a different scope then #1081. let's focus on the original request.

lmcbout commented:
@AlexTugarev 
I will bring back the overload, and   look at your latest comment.
But just as an example, here a link why we should be using less overload, just read the first paragraph of https://github.com/Microsoft/TypeScript/issues/12041 . It says: 
"Today's function overloading in TypeScript is a compromise which is the worst of both worlds of dynamic, non-overloaded language like JS, and static, overloaded language such as TS."

There are also other links talking on the subject, but stop here.
I personnally think  regrouping all optional parameters within the same API would reduce the complexity to maintain the code in a long run.

lmcbout commented:
@AlexTugarev I just created a new version for the auto-dismiss with the function overload. I think I covered all your previous comments. If there is something missing, I would be happy to get your version for the fix, I will be happy to merge it if ever you find something missing form previous version.

AlexTugarev commented:
@lmcbout, could you please reopen this?

AlexTugarev commented:
@lmcbout 
> But just as an example, here a link why we should be using less overload, just read the first paragraph of Microsoft/TypeScript#12041 . It says:
"Today's function overloading in TypeScript is a compromise which is the worst of both worlds of dynamic, non-overloaded language like JS, and static, overloaded language such as TS."

That sounds like Shlomibo tried to make a case to change overloads in TS, which was declined. I do not see a reason to stop using it, just because people wish it would be extended.

> I personnally think regrouping all optional parameters within the same API would reduce the complexity to maintain the code in a long run.

Couldn't agree more. It's the point about making the breaking change right now.
